### PR TITLE
Issue538 seasurfaceheight

### LINKED ISSUE #538: Gradient of sea-surface height
The momentum equation solver for all rheologies is missing the sea-surface-tilt term. We need

- [x] Routine to calculate the gradient of this field
- [x] Passing the sea-surface field from the ocean inputs to the dynamics

### PR DESCRIPTION
**Pull request by einola**
# Add sea-surface height
Fixes #538 

### Task List
- [x] Defined the tests that specify a complete and functioning change (*It may help to create a [design specification & test specification](../../../wiki/Specification-Template)*)
- [x] Implemented the source code change that satisfies the tests
- [ ] Documented the feature by providing worked example
- [ ] Updated the README or other documentation
- [x] Completed the pre-Request checklist below

---
# Change Description

This PR adds the contribution of the sea-surface-height gradient to both the mEVP and BBM codes (through VPCGDynamicsKernel and BrittleDynamicsKernel). SSH is read in from IOceanBoundary and passed to the dynamical core. It can also be output through ConfigOutput as "ssh".

---
# Test Description

No unit test was implemented for this feature, and no quantitative test exists. Qualitative testing shows that setting a fixed sea-surface-height gradient gives drift speed close to back-of-the-envelope estimates (13 cm/s modelled, 20 cm/s from back-of-the-envelope).

---
# Documentation Impact

None

---
# Other Details

None

---
### Pre-Request Checklist

- [x] The requirements of this pull request are fully captured in an issue or design specification and are linked and summarised in the description of this PR
- [x] No new warnings are generated
- [x] The documentation has been updated (or an issue has been created to track the corresponding change)
- [x] Methods and Tests are commented such that they can be understood without having to obtain additional context
- [x] This PR/Issue is labelled as a bug/feature/enhancement/breaking change
- [x] File dates have been updated to reflect modification date
- [x] This change conforms to the conventions described in the README

### CODE REVIEW

File: physics/src/modules/include/IOceanBoundary.hpp [hunk 1]
```diff
@@ -86,6 +89,7 @@ class IOceanBoundary : public ModelComponent {
     HField cpml; // Heat capacity of the mixed layer, J K⁻¹ m²
     UField u; // x(east)-ward ocean current, m s⁻¹
     VField v; // y(north)-ward ocean current, m s⁻¹
+    VField ssh; // sea surface height, m
```
Comment (timspainNERSC) [inline physics/src/modules/include/IOceanBoundary.hpp]: This should be an `HField`, I believe.

**Review by timspainNERSC — CHANGES_REQUESTED** [review]
Just one small change. I am assuming that the sea-surface gradient calculation code is correct.

**Comment by winzerle** [comment]
It should be a variable like H and A, so HField appears good. 


* --------------------------------------------
| Prof. Dr. Thomas Richter
| Numerische Mathematik in den Anwendungen
| Otto-von-Guericke Universität Magdeburg
| numerics.ovgu.de
| +49 391 67-57162  +49 176 615 612 26 (mobil)

File: physics/src/modules/include/IOceanBoundary.hpp (hunk 1 above)
Comment (einola) [reply physics/src/modules/include/IOceanBoundary.hpp]: Yes! Good catch.

File: core/src/modules/include/ProtectedArrayNames.ipp [hunk 2]
```diff
@@ -8,37 +8,38 @@
...
+    { "u", "ICE_U" }, // x(east)-ward ice velocity, m s⁻¹
+    { "v", "ICE_V" }, // y(north)-ward ice velocity, m s⁻¹
+    { "sst_slab", "SLAB_SST" }, // Slab ocean surface temperature ˚C
+    { "sss_slab", "SLAB_SSS" }, // Slab ocean surface salinity PSU
+    { "qdw", "SLAB_QDW" }, // Slab ocean temperature nudging heat flux, W m⁻²
+    { "fdw", "SLAB_FDW" }, // Slab ocean salinity nudging water flux, kg s⁻¹ m⁻²
 { "ssh", "SSH" }, // Slab ocean salinity nudging water flux, kg s⁻¹ m⁻²
-
+{ "taux", "IO_STRESS_U" }, // Ice-ocean stress x(east) direction, Pa
+{ "tauy", "IO_STRESS_V" }, // Ice-ocean stress x(east) direction, Pa
```
Comment (timspainNERSC) [inline core/src/modules/include/ProtectedArrayNames.ipp]: Some, but not all, of these are indented. Can they at least be consistent?

File: physics/src/modules/include/IOceanBoundary.hpp [hunk 3]
```diff
@@ -13,12 +13,12 @@
...
-constexpr TextTag MLD = "MLD"; // Mixed layer or slab ocean depth m
-constexpr TextTag OCEAN_U = "U"; // x(east)-ward ocean current m s⁻¹
-constexpr TextTag OCEAN_V = "V"; // y(north)-ward ocean current m s⁻¹
-constexpr TextTag SSH = "SSH"; // sea surface height, m
+    constexpr TextTag SST = "SST"; // sea surface temperature ˚C
+    constexpr TextTag SSS = "SSS"; // sea surface salinity PSU
+    constexpr TextTag MLD = "MLD"; // Mixed layer or slab ocean depth m
+    constexpr TextTag OCEAN_U = "U"; // x(east)-ward ocean current m s⁻¹
+    constexpr TextTag OCEAN_V = "V"; // y(north)-ward ocean current m s⁻¹
+    constexpr TextTag SSH = "SSH"; // sea surface height, m
```
Comment (timspainNERSC) [inline physics/src/modules/include/IOceanBoundary.hpp]: If this is `clang-format`, then it is fine, but I thought that a `namespace` block didn't typically gain a level of indentation.

**Review by timspainNERSC — APPROVED** [review]
A couple of formatting changes, but if they are what `clang-format` demands, then they shall stand as they are.

File: core/src/modules/include/ProtectedArrayNames.ipp (hunk 2 above)
Comment (einola) [reply core/src/modules/include/ProtectedArrayNames.ipp]: Yes, this is weird. I removed the indentation. I'm not sure if clang-format knows what to do with this file.

File: physics/src/modules/include/IOceanBoundary.hpp (hunk 3 above)
Comment (einola) [reply physics/src/modules/include/IOceanBoundary.hpp]: They do gain a level of indentation when nested in other namespaces (see [here](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#namespaceindentation) - we use NamespaceIndentation: Inner)

**Pull request merged**